crates/sc-consensus-subspace/src/archiver.rs

```diff
@@ -952,17 +952,19 @@ where
         return Ok(true);
     }
 
-    let maybe_block_hash = client.hash(block_number_to_archive)?;
-
-    let Some(block_hash) = maybe_block_hash else {
-        warn!("Archiver detected an error: older block by number must always exist. ");
-
+    if block_number_to_archive > *best_archived_block_number + 1u32.into() {
```

Comment:
This is technically correct, but it is correct incidentally. The invariant you're actually checking here is whether the block we're trying to archive follows best archived block exactly, so better self-documenting version of the code is this:
```suggestion
    if block_number_to_archive != *best_archived_block_number + 1u32.into() {
```

^ clearly says that we're doing for `block_number_to_archive == best_archived_block_number + 1` and if that is not the case we need to do something else.

Also to make it look more like a state machine I'd probably change API to return `Option<(best_archived_block_number, best_archived_block_hash)>`, which in case `None` is returned would mean archiver reinitialization is necessary.